skbio/metadata/_interval.py

```diff
@@ -0,0 +1,312 @@
+
+
+class Interval():
+    '''Store the metadata of a sequence interval.
+
+    Parameters
+    ----------
+    intervals : list of tuple of ints
+        List of tuples representing start and end coordinates.
+    boundaries : list of tuple of bool
+        List of tuples, representing the openness of each interval.
+    metadata : dict
+        Dictionary of attributes storing information of the feature
+        such as `strand`, `gene_name` or `product`.
+    interval_metadata : object
+        A reference to the `IntervalMetadata` object that this
+        interval is associated to.
+    '''
+    def __init__(self, intervals=None, boundaries=None,
+                 metadata=None, interval_metadata=None):
+        iv = []
+        for interval in intervals:
+            iv.append(_polish_interval(interval))
+
+        if boundaries is not None:
+            self.boundaries = boundaries
+        else:
+            self.boundaries = []
+
+        if metadata is not None:
+            self.metadata = metadata
+        else:
+            self.metadata = {}
+
+        self._interval_metadata = interval_metadata
+
+        if intervals is not None:
+            self.intervals = iv
+        else:
+            self.intervals = []
+
+    def __getitem__(self, key):
+        return self.metadata[key]
+
+    def __setitem__(self, key, val):
+        self.metadata[key] = val
+
+    # This is required for creating unique sets of intervals
+    def __hash__(self):
+        return hash(tuple(sorted(self.metadata.items()) +
+                          self.intervals +
+                          self.boundaries))
+
+    def __lt__(self, other):
+        return self.intervals < other.intervals
+
+    def __gt__(self, other):
+        return self.intervals > other.intervals
+
+    def __le__(self, other):
+        return self.intervals <= other.intervals
+
+    def __ge__(self, other):
+        return self.intervals >= other.intervals
+
+    def __eq__(self, other):
+        return ((self.metadata == other.metadata) and
+                (self.intervals == other.intervals) and
+                (self.boundaries == other.boundaries))
+
+    def __contains__(self, key):
+        return key in self.metadata
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __repr__(self):
+        return ''.join([
+            "Interval("
+            "intervals=" + str(self.intervals),
+            ", metadata=" + str(self.metadata),
+            ")"])
+
+    def drop(self):
+        self._interval_metadata.drop(intervals=self.intervals,
+                                     boundaries=self.boundaries,
+                                     metadata=self.metadata)
+        self.boundaries = None
+        self.intervals = None
+        self._interval_metadata = None
+
+    @property
+    def intervals(self):
+        return self._intervals
+
+    @intervals.setter
+    def intervals(self, value):
+        self._intervals = value
```

Comment:
I think there should be some validation/normalization for `value`. Right now it can be anything and the user won't get an error until the tree gets rebuilt, which doesn't happen right away.